internal/services/communication/communication_service_email_domain_association_resource.go

```diff
@@ -0,0 +1,340 @@
+			}
+
+			existingCommunicationService, err := client.Get(ctx, *communicationServiceId)
+			if err != nil && !response.WasNotFound(existingCommunicationService.HttpResponse) {
+				return fmt.Errorf("checking for the presence of existing  %s: %+v", communicationServiceId, err)
+			}
+
+			if response.WasNotFound(existingCommunicationService.HttpResponse) {
+				return fmt.Errorf("%s was not found", communicationServiceId)
+			}
+
+			if existingCommunicationService.Model == nil {
+				return fmt.Errorf("model for %s was nil", communicationServiceId)
+			}
+
+			if existingCommunicationService.Model.Properties == nil {
+				return fmt.Errorf("properties for %s was nil", communicationServiceId)
+			}
+
+			domainList := make([]string, 0)
+			if existingDomainList := existingCommunicationService.Model.Properties.LinkedDomains; existingDomainList != nil {
+				domainList = pointer.From(existingDomainList)
+			}
+
+			id := commonids.NewCompositeResourceID(communicationServiceId, eMailServiceDomainId)
+
+			for _, v := range domainList {
+				tmpID, tmpErr := domains.ParseDomainIDInsensitively(v)
+				if tmpErr != nil {
+					return fmt.Errorf("parsing domain ID %q from LinkedDomains for %s: %+v", v, communicationServiceId, err)
+				}
+
+				if strings.EqualFold(eMailServiceDomainId.ID(), tmpID.ID()) {
+					return metadata.ResourceRequiresImport(r.ResourceType(), id)
+				}
+			}
+
+			domainList = append(domainList, eMailServiceDomainId.ID())
+
+			input := communicationservices.CommunicationServiceResourceUpdate{
+				Properties: &communicationservices.CommunicationServiceUpdateProperties{
+					LinkedDomains: pointer.To(domainList),
+				},
+			}
+
+			if _, err = client.Update(ctx, *communicationServiceId, input); err != nil {
+				return fmt.Errorf("updating %s: %+v", *communicationServiceId, err)
+			}
+
+			metadata.SetID(id)
+
+			return nil
+		},
+	}
+}
+
+func (EmailDomainAssociationResource) Read() sdk.ResourceFunc {
+	return sdk.ResourceFunc{
+		Timeout: 5 * time.Minute,
+		Func: func(ctx context.Context, metadata sdk.ResourceMetaData) error {
+			client := metadata.Client.Communication.ServiceClient
+			domainClient := metadata.Client.Communication.DomainClient
+
+			id, err := commonids.ParseCompositeResourceID(metadata.ResourceData.Id(), &communicationservices.CommunicationServiceId{}, &domains.DomainId{})
+			if err != nil {
+				return err
+			}
+
+			state := EmailDomainAssociationResourceModel{}
+			state.CommunicationServiceId = id.First.ID()
+			state.EMailServiceDomainId = id.Second.ID()
+
+			communicationServiceId, err := communicationservices.ParseCommunicationServiceID(state.CommunicationServiceId)
+			if err != nil {
+				return fmt.Errorf("parsing Communication Service ID: %w", err)
+			}
+
+			eMailServiceDomainId, err := domains.ParseDomainID(state.EMailServiceDomainId)
+			if err != nil {
+				return fmt.Errorf("parsing EMail Service Domain ID: %w", err)
+			}
```

Comment:
Sorry, looks like I missed a couple of unnecessarily wrapped errors
```suggestion
			communicationServiceId, err := communicationservices.ParseCommunicationServiceID(state.CommunicationServiceId)
			if err != nil {
				return err
			}

			eMailServiceDomainId, err := domains.ParseDomainID(state.EMailServiceDomainId)
			if err != nil {
				return err
			}
```